experimental/InjectiveResolutions/src/InjectiveResolutions.jl

```diff
@@ -0,0 +1,1098 @@
+export monoid_algebra_ideal
+export faces
+export hyperplanes
+
+export irreducible_resolution
+export irreducible_decomposition
+
+export injective_resolution
+
+export local_cohomology
+export local_cohomology_all
+export zeroth_local_cohomology
+
+export MonoidAlgebra
+export MonoidAlgebraIdeal
+export MonoidAlgebraElem
+
+#########################
+# some composite types
+#########################
+
+include("MonoidAlgebra.jl")
+
+struct IndecInj #indecomposable injective
+  face::FaceQ
+  vector::Vector{Int}
+end
+mutable struct InjMod #ZZ^d i-graded injective module over monoid algebra
+  monoid_algebra::MonoidAlgebra
+  indec_injectives::Vector{IndecInj}
+  Q_graded_part::Union{SubquoModule,Nothing}
+
+  function InjMod(A::MonoidAlgebra,I::Vector{IndecInj})
+    return new(A,I,nothing)
+  end
+end
+
+function Q_graded_part(I::InjMod)
+  if I.Q_graded_part === nothing
+    I.Q_graded_part = compute_Q_graded_part(I)
+  end
+  return I.Q_graded_part
+end
+
+function compute_Q_graded_part(I::InjMod)
+  kQ = I.monoid_algebra
+  irreducible_ideals = [_get_irreducible_ideal(kQ, J) for J in I.indec_injectives]
+  irreducible_comp = [quotient_ring_as_module(Ji) for Ji in irreducible_ideals]
+  return direct_sum(irreducible_comp...,task=:none)
+end
+struct IrrRes # irreducible resolution (including all computed data and the cochain complex)
+  mod::SubquoModule
+  irr_sums::Vector{InjMod}
+  cochain_maps::Vector{SubQuoHom}
+  cochain_complex::ComplexOfMorphisms # if sequence not exact return trivial cochain_complex (M0 -> M0)
+end
+
+#direct sum of two injective modules over the same monoid algebra
+function +(I::InjMod, J::InjMod)
+  @req I.monoid_algebra == J.monoid_algebra "monoid algebras not the same"
+  return InjMod(I.monoid_algebra, vcat(I.indec_injectives, J.indec_injectives))
+end
+
+struct InjRes #ZZ^d-graded injective resolution
+  mod::SubquoModule
+  inj_mods::Vector{InjMod}
+  cochain_maps::Vector{MatElem}
+  upto::Int
+  Q_graded_part::IrrRes
+  shift::Vector{Int} #not needed
+end
+
+function Base.show(io::IO,J::InjMod)
+  print(
+    io, "injective module given by direct sum of ", length(J.indec_injectives)," indecomposable injectives"
+  )
+end
+
+function Base.show(io::IO, ::MIME"text/plain", J::InjMod)
+  println(io, "injective module given by direct sum of indecomposable injectives")
+  for Ji in J.indec_injectives
+    println(io, "  k{", Ji.vector, " + F - Q}, where p_F = ", Ji.face.prime)
+  end
+  print(
+    io,
+    "over ",
+    J.monoid_algebra
+  )
+end
+
+function Base.show(io::IO, ::MIME"text/plain", res::InjRes)
+  println(io, "injective resolution ")
+  println(io, "  ", join(["J^$i" for i in 0:res.upto], " -> "))
+  println(io, "where ")
+  for i in eachindex(res.inj_mods)
+    j = i-1
+    println(io, " J^$j = direct sum of")
+    for Ji in res.inj_mods[i].indec_injectives
+      println(io, "    k{", Ji.vector, " + F - Q}, where p_F = ", Ji.face.prime)
+    end
+  end
+  println(io, "of ", res.mod)
+  println(
```

Comment:
```suggestion
  print(
```
and similar elsewhere: your `show` methods should never end with a `println` / a trailing newline.